internal/target/service_list.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package target
+
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/refreshtoken"
+)
+
+// This function is a callback passed down from the application service layer
+// used to filter out protobuf targets that don't match any user-supplied filter.
+type ListFilterFunc func(Target) (bool, error)
+
+// List lists targets according to the page size,
+// filtering out entries that do not
+// pass the filter item fn. It returns a new refresh token
+// based on the grants hash and the returned targets.
+func List(
+	ctx context.Context,
+	repo *Repository,
+	grantsHash []byte,
+	pageSize int,
+	filterItemFn ListFilterFunc,
+) (*ListResponse, error) {
+	const op = "target.List"
+
+	limit := pageSize + 1
+	opts := []Option{
+		WithLimit(limit),
+	}
+
+	targets := make([]Target, 0, limit)
+dbLoop:
+	for {
+		// Request another page from the DB until we fill the final items
+		page, err := repo.ListTargets(ctx, opts...)
```

Comment:
With this change, I think we should make the `ListTargets` method on the repository un-exported.

Reply:
Good idea

Reply:
I had a quick look at this and unfortunately it means that it's very difficult to perform database testing of this method. Since all concrete implementations of targets exist in subpackages that import this package, it is impossible to use these types in a package testing this function (if it is unexported). I'm thus tempted to leave this exported just for ease of testing. We had the same problem with ListDeletedIds and EstimatedCount but those were easier to mock. I'd like to use the real DB for the list targets test since it relies heavily on the query for ordering and so on.

Reply:
You could add something to [exports_test](https://github.com/hashicorp/boundary/blob/main/internal/target/exports_test.go) to expose the method for tests, while leaving it un-exported outside of tests.

Reply:
🤯 I didn't know we had a pattern like this, thank you!